magpie/utils.py

```diff
@@ -279,6 +281,53 @@ def convert_response(response):
     return pyramid_response
 
 
+def get_authenticate_headers(request, error_type="invalid_token"):
+    # type: (Request, Str) -> Optional[HeadersType]
+    """
+    Obtains all required headers by 401 responses based on executed :paramref:`request`.
+
+    :param request: request that was sent to attempt authentication or access which must respond with Unauthorized.
+    :param error_type: additional detail of the cause of error, one of (invalid_token, invalid_token
+    """
+    # FIXME: support other authentication methods (JWT, HTTP, Basic, Bearer Token, etc.)
+    #        in such case, must resolve specified challenge method according to request if provided
+    #        (https://github.com/Ouranosinc/Magpie/issues/255)
+    # Generic Auth: https://tools.ietf.org/html/rfc7235#section-2.1  (section for schema of 'WWW-Authenticate')
+    # Basic/Digest: https://tools.ietf.org/html/rfc2617
+    # Bearer Token: https://tools.ietf.org/html/rfc6750
+    # Cookie Token: https://tools.ietf.org/id/draft-broyer-http-cookie-auth-00.html#anchor1
+
+    from magpie.api.schemas import SigninAPI  # pylint: disable=C0415
+
+    # avoid adding headers when explicitly requested
+    #   https://stackoverflow.com/questions/9859627
+    #   https://stackoverflow.com/questions/86105
+    if get_header("X-Requested-With", request.headers) == "XMLHttpRequest":
+        return None
+
+    # select error type: https://tools.ietf.org/html/rfc6750#section-3.1
+    if error_type not in ["invalid_token", "invalid_token", "insufficient_scope"]:
+        error_type = "invalid_token"
+    cookie_name = get_constant("MAGPIE_COOKIE_NAME", request)
+    magpie_url = get_magpie_url(request)
+    signin_url = "{}{}".format(magpie_url, SigninAPI.path)
+    login_url = "{}/ui/login".format(magpie_url)
```

Comment:
Do we have a LoginAPI.path variable (like the SigninAPI.path found in the preceding line)?

Reply:
Sadly no. All UI paths don't have a service as they are not part of the OpenAPI.